### PR TITLE
Use sealed securityContext in SecureChannel

### PR DESCRIPTION
**Pull request by benaadams**
Allows the virtual call in IsAuthenticated to become direct and inline and the method to be 2.5x faster (own time 35% -> 100%, Time 160ms -> 60ms)

Pre

image

Post

image

Pre is preview8; Post also includes the improvements from https://github.com/dotnet/corefx/pull/39950; though that didn't effect IsAuthenticated.

Also now tailcalls into `DecryptMessage` which is why `SecureChannel.Decrypt` disappears from the trace

/cc @davidsh @wfurt @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
> Time 160ms -> 60ms

So it saves 100ms out of how many?

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs [hunk 1]
```diff
@@ -25,6 +26,7 @@ internal struct CredHandle
 
             public bool IsZero
             {
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: Why is this needed?

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs [hunk 2]
```diff
@@ -33,6 +34,7 @@ protected SafeDeleteContext() : base(IntPtr.Zero, true)
 
         public override bool IsInvalid
         {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: Why is this needed?

File: src/Common/src/System/Net/Security/NegotiateStreamPal.Windows.cs [hunk 3]
```diff
@@ -109,17 +109,19 @@ internal static string QueryContextAuthenticationPackage(SafeDeleteContext secur
             var outSecurityBuffer = new SecurityBuffer(resultBlob, SecurityBufferType.SECBUFFER_TOKEN);
 
             Interop.SspiCli.ContextFlags outContextFlags = Interop.SspiCli.ContextFlags.Zero;
+            // There is only one SafeDeleteContext type on Windows which is SafeDeleteSslContext so this cast is safe.
```
Comment (stephentoub) [inline src/Common/src/System/Net/Security/NegotiateStreamPal.Windows.cs]: Is there more than one on Unix?

Comment (benaadams) [reply src/Common/src/System/Net/Security/NegotiateStreamPal.Windows.cs]: 1 on macOS, 1 on Windows, 2 on Unix (negotiate stream uses a different derived one `SafeDeleteNegoContext`) which is why I assume `SafeDeleteContext` is abstract. SslStream uses `SafeDeleteContext` on all platforms so they all benefit.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 4]
```diff
@@ -139,6 +140,7 @@ internal SslConnectionInfo ConnectionInfo
 
         internal bool IsValidContext
         {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Why is this needed?

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 5]
```diff
@@ -351,8 +353,7 @@ private bool AcquireClientCredentials(ref byte[] thumbPrint)
                 X509Certificate2 remoteCert = null;
                 try
                 {
-                    X509Certificate2Collection dummyCollection;
-                    remoteCert = CertificateValidationPal.GetRemoteCertificate(_securityContext, out dummyCollection);
+                    remoteCert = CertificateValidationPal.GetRemoteCertificate(_securityContext, out X509Certificate2Collection dummyCollection);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Where is `dummyCollection` used?  It seems like it's not, in which case it could use a discard.  However, do we have an existing issue here around cleaning up anything in this collection?  @bartonjs, should we be disposing of the contents of the collection?

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 6]
```diff
@@ -649,8 +650,10 @@ private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
             }
             else if (_sslAuthenticationOptions.CertSelectionDelegate != null)
             {
-                X509CertificateCollection tempCollection = new X509CertificateCollection();
-                tempCollection.Add(_sslAuthenticationOptions.ServerCertificate);
+                X509CertificateCollection tempCollection = new X509CertificateCollection
+                {
+                    _sslAuthenticationOptions.ServerCertificate
+                };
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Not sure this is an improvement.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 7]
```diff
@@ -452,7 +452,7 @@ public Task AuthenticateAsServerAsync(SslServerAuthenticationOptions sslServerAu
                 this);
         #endregion
 
-        public override bool IsAuthenticated => _context != null && _context.IsValidContext && _exception == null && _handshakeCompleted;
+        public override bool IsAuthenticated => _exception == null && _handshakeCompleted && (_context?.IsValidContext ?? false);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Why reorder this?

**Comment by benaadams** [comment]
> So it saves 100ms out of how many?

Depends on the size of what's being encrypted; small websocket messages it can be significant, larger chunks less so.

The main underlying issue is its x3 call count to reads and writes (which are virtuals so can't be optimized away https://github.com/dotnet/corefx/pull/39986#issuecomment-518256995)

image

For smaller message sizes it currently it spends more time checking IsAuthenticated than doing Encryption.

image

IsAuthenticated is currently 2 indirect calls, 2 non-inlines; this takes it to 1 indirect call.

**Comment by stephentoub** [comment]
> For smaller message sizes it currently it spends more time checking IsAuthenticated than doing Encryption.

Surprising.  Thanks for the data.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: It was initially to tailcall into `IsValidContext->_securityContext.IsInvalid` when I wasn't sure I could get it to devirtualize.

Can revert if preferred?

File: src/Common/src/System/Net/NTAuthentication.Common.cs [hunk 8]
```diff
@@ -203,8 +202,7 @@ internal string GetOutgoingBlob(string incomingBlob)
 
         internal byte[] GetOutgoingBlob(byte[] incomingBlob, bool thrownOnError)
         {
-            SecurityStatusPal statusCode;
-            return GetOutgoingBlob(incomingBlob, thrownOnError, out statusCode);
+            return GetOutgoingBlob(incomingBlob, thrownOnError, out _);
```
Comment (bartonjs) [inline src/Common/src/System/Net/NTAuthentication.Common.cs]: Unless there's something I don't know about `out _`, the edits to this file are style without substance, and thus this file should be dropped from the PR.

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 5 above)
Comment (bartonjs) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Yeah, if it's non-null then everything in it should be Disposed.  The only other caller that I see of GetRemoteCertificate already disposes it (this one is probably a lot less used).

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (bartonjs) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: If there's a benefit, add a comment for how someone can make sure they don't undo it.  If there's no benefit, then this file should be completely reverted.

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 1 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: Doesn't inline

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs (hunk 2 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: Doesn't inline

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 4 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Doesn't inline

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 6 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Was just following the `.editorconfig` https://github.com/dotnet/corefx/blob/963e0d64bd7a69ba4b8d4d933bd56d0f2a5a8913/.editorconfig#L89 will revert

Comment (stephentoub) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Thanks.  We need to revise that config file.  But it's also appreciated to avoid applying any refactorings for style in PRs intended for other purposes; makes it hard to find the signal in the noise.

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 7 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Removed; with the devirtualization it no longer provides benefit, it was a step on the way

File: src/Common/src/System/Net/NTAuthentication.Common.cs (hunk 8 above)
Comment (benaadams) [reply src/Common/src/System/Net/NTAuthentication.Common.cs]: Reverted

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs (hunk 2 above)
Comment (bartonjs) [reply src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: > Doesn't inline

... and?  Is that a 0.00000001% improvement, or does it actually move a needle on anything beyond a microbenchmark?

The JIT has heuristics for determining when to inline.  If there's a general pattern that it misses or scores "incorrectly" then we should solve it in the JIT and leave AggressiveInlining for those cases where a) it looks like, but is provably different, from a thing the JIT says no to and b) it actually moves the needle on real code.

**Comment by benaadams** [comment]
`[MethodImpl(MethodImplOptions.AggressiveInlining)]`

> > Why is this needed?

> Doesn't inline

With it inlining, do still end up with some slightly odd bool handling logic in the asm ([simplified sharplab.io sample](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNXjxldskhrgxQOYDAyGwAJgAlsXK9JoBvGg3cMxUXguwYYDACSPAAKxgxWAO4AMhCRMhrUHp7evv5BoeFRAKpiYgmaSR7EAMwMwBAQBoG4AFoyEG4erkXJ7gDaALIwGAAWEFaB4pIAFN19A0NikgDyYnwQXLgsAIIA5muwuLg+MMGS2tprAJQAuk1tDGs9DM5MAOwRMXEyDAC8bxkYYVAs9VAQBgAMiBT1y+Sg70+wW+xj+DTUDAAvhdkTQUdQaLgYPoYFYmKQGLJsAAzGAAERg0n8Ih0CHMICYrE4PH4ghhDTE8igShUy2JZLsDic1BayW0/igegMljxQscAQA+r17ArEslUV4IP4zHiiaSKVSejBaf56SNjgxGcA8DARjCfvCAWgGMYODBjqixZcGMrVdJ3gwuDBIhZrPLpBb1R4MRrWkwyhAFDJvFYAhUqkFcMFfAcrF7Uckuj1+oNhmMS5NhnMFktVhstjtk/tDlwTud48lrhhC81e+5iI8akJJBBsfjCIRfSrhey6gj+2jOzHCsktTr/PikyneGnypUDGwjbaI3bPfHvZc+gDQ8HQwA5bVTaQCHRWACiCBU814iyjqIxDEaDIIkYDADhYCEGdg0kFxUQFQ1qRNRYzXMRVsXA7wMAAT1NGB6WjS5UQlXR9H3TMagANX0Xc8PpAtlw6cZS2fCsJjLaYa1/Ot1k2VQmz2LgDi4I4zkXbtF0vH0B0eABCEZ0LAiDeBwujzA+IMOEkAxJ19DDlNUlD8IwFgahzGirGOQi2ljFdqCA6hsGAIwoGwMwCSJXBJFkYwcXwOD43kTCTWgqlfUgOkMGsz9v1rX18Ni7jrIzAw/QcXAVQAa2Q4ZjSsayLgASFKBglCgDAODIlKsxWDg+hgVkwD8PU3gAPnCoz6QYWTPi4LSDBBDrItM3BqLzNTgVBRUEpgH9FihTTtMm6dVQy7BspEXLN0SJEgA===))
```asm
SslStream.get_IsAuthenticated()
    L0000: cmp qword [rcx+0x8], 0x0
    L0005: jz L0058
    L0007: mov rax, [rcx+0x8]
    L000b: mov rax, [rax+0x8]
    L000f: test rax, rax
    L0012: jz L0058
    L0014: mov edx, [rax]
    L0016: test byte [rax+0x10], 0x1
    L001a: jnz L003b
    L001c: add rax, 0x18
    L0020: mov rdx, [rax]
    L0023: test rdx, rdx         <
    L0026: jnz L0037             <
    L0028: mov rax, [rax+0x8]    <
    L002c: test rax, rax         <
    L002f: setz al               <
    L0032: movzx eax, al         <
    L0035: jmp L0039             <
    L0037: xor eax, eax          <
    L0039: jmp L0040             <
    L003b: mov eax, 0x1          <
    L0040: test eax, eax         <
    L0042: setz al               <
    L0045: movzx eax, al         <
    L0048: test eax, eax         <
    L004a: jz L0058              <
    L004c: cmp qword [rcx+0x10], 0x0
    L0051: jnz L0058
    L0053: movzx eax, byte [rcx+0x18]
    L0057: ret
    L0058: xor eax, eax
    L005a: ret
```
So if L002c was hit it then follows this sequence, which everything between the first test and last jump test is redundant?
```asm
    L002c: test rax, rax         
    L002f: setz al               
    L0032: movzx eax, al
    L0040: test eax, eax  
    L0042: setz al      
    L0045: movzx eax, al
    L0048: test eax, eax
    L004a: jz L0058  
```

/cc @mikedn @AndyAyersMS

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs (hunk 2 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: > ... and? Is that a 0.00000001% improvement

Not inlining pushes IsAuthenticated back to being slower than the crypto for small messages; due to the high ratio of its call count and bools having messy asm

image


>  does it actually move a needle on anything beyond a microbenchmark?

I'm not microbenchmarking; this is from our real application which is entire raison d'être is very high speed websocket messages.

>  If there's a general pattern that it misses or scores "incorrectly" then we should solve it in the JIT 

I'd prefer the JIT's handling of bools to be improved as when inlined it could produce far better asm than it currently does. Currently it does a test, then materilizes that test to a register to pass it back as the return; which it then does a test on to get the bool back, which the vestiges of which still remain after inlining which should be optimzed away https://github.com/dotnet/corefx/pull/40265#issuecomment-520901302

> a) it looks like, but is provably different, from a thing the JIT says no to

And this is what the traces are showing; maybe it doesn't like bools for the reason above. The common path is to execute almost all of the tests as the common state is valid and since anything can make it invalid it needs to check everything so they aren't conditional calls in the common path.

> b) it actually moves the needle on real code.

This is real code and every trace in this PR has been from real code; and while WebSockets and Kestrel's streams are bothering me, SslStream is my main nemesis:

image

And this is one of many steps to address its issues

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 9]
```diff
@@ -908,7 +909,8 @@ internal void ProcessHandshakeSuccess()
         --*/
         internal SecurityStatusPal Encrypt(ReadOnlyMemory<byte> buffer, ref byte[] output, out int resultSize)
         {
-            if (NetEventSource.IsEnabled)
+            bool logEnabled = NetEventSource.IsEnabled;
+            if (logEnabled)
```
Comment (davidsh) [inline src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Why are we changing all the 'NetEventSource.IsEnabled' to 'logEnabled'? Is it really that much faster. 

We already a coding pattern everywhere else in networking code like this:

```c#
if (NetEventSource.IsEnabled) NetEventSource.method
```

Are we going to change it everywhere else in the codebase?

This change diminishes readability and the benefit in potential performance is questionable.

**Comment by benaadams** [comment]
If I change them all to ternary operators then the asm becomes better [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNXjxldskhrgxQOYDAyGwAJgAlsXK9JoBvGg3cMxUXguwYYDACSPAAKxgxWAO4AMhCRMhrUHp7evv5BoeFRAKpiYgmaSR7EAMwMwBAQBoG4AFoyEG4erkXJ7gDaALIwGAAWEFaB4pIAFN19A0NikgDyYnwQXLgsAIIA5muwuLg+MMGS2tprAJQAuk1tDGs9DM5MAOwRMXEyDAC8bxkYYVAs9VAQBgAMiBT1y+Sg70+wW+xj+DQYAH4GMYOAEQAwAGb6XAwNQMAC+F0JNCJ1BouP0MCsTFIDFk2ExMAAIjBpP4RDoEOYMcx2Nw+AIWDCGmJ5FAlCplgymXYHE5qC1ktp/FA9AZLNS5Y4AgB9Xr2HWJZLErwQfxman0xkstk9GCc/zckbHBgY4B4GAjGE/eEAtAokwwY7EpWXBj6w3Sd4MLgwSIWaza6Qu40eMkm1pMMoQBQybxWAIVKpBXDBXwHKyh4nJLo9fqDYZjeuTYZzBZLVYbLY7PP7Q5cE7nLPJa4YGvNCfuACQxEeNSEkgguJphEIEYN8sENX+gORqPRWJxeKnJKzZNNAIt/hpufzvEL5UqBjYds9ya9IazYcufQBCbjBMADlzSmaQBB0KwAFEEBUeZeEWVNiTJC9aDpeQwA4WAhE3ONJBcYkZVtdkHUWJ1zF1XFMO8DAAE9HRgbk00uYkVV0fQnxLGoADV9AfBjuWrEcPDrCZG2mZsxLA9sEM7dZNlUXs9i4A4uCOM5TzHU8f3Ddw5wYEZKJgajeDogTzA+WMOEkAw1wjKisNM+iyMYjBhTLLgKwfJEj0kXE3UDNFjmYtoM3TUlCmwYAjCgbAzFpelcEkWRjBgbB8AIrMMKwh1cLZCNIC5DAQpguCOwjRiytkkLiwMSMHFwA0AGtSOGe0rBCi5ZzKJQoAwDgONq0sVg4PoYB4ZQ/CtN4AD4Cpc7kGAAQk+LhrIMEF5qK9zeMrczgVBXVKpgeDFihKybIOjdDUa7AWpENqb0SAkgA=)
    
```diff
SslStream.get_IsAuthenticated()
    cmp qword [rcx+0x8], 0x0
-   jz L0058
+   jz L003d
    mov rax, [rcx+0x8]
    mov rax, [rax+0x8]
    test rax, rax
-   jz L0058
+   jz L003d
    mov edx, [rax]
    test byte [rax+0x10], 0x1
-   jnz L003b
+   jnz L003d
    add rax, 0x18
    mov rdx, [rax]
    test rdx, rdx         
-   jnz L0037     
+   jnz L0031
    mov rax, [rax+0x8]    
    test rax, rax    
-   setz al               
-   movzx eax, al         
-   jmp L0039             
-   xor eax, eax          
-   jmp L0040             
-   mov eax, 0x1          
-   test eax, eax         
-   setz al               
-   movzx eax, al         
-   test eax, eax         
-   jz L0058     
+   jz L003d              
    cmp qword [rcx+0x10], 0x0
-   jnz L0058
+   jnz L003d
    movzx eax, byte [rcx+0x18]
    ret
    xor eax, eax
    ret
```
but I know @mikedn doesn't like that 😉

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 9 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Happy to revert; only adds 1 extra mov to the tests.

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 1 above)
Comment (stephentoub) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: > Doesn't inline

I'm surprised that `return dwLower == IntPtr.Zero && dwUpper == IntPtr.Zero;` wouldn't inline.  What was the reasoning it gave?
cc: @AndyAyersMS

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs (hunk 2 above)
Comment (bartonjs) [reply src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: And, just to be sure, with the AggressiveInline on IsAuthenticated it actually removes 250ms from this test (presumably the overhead of calling a function 96,000 times) (and plays accounting games with the other 110ms)?

'Cuz that I'd admit is moving the needle :wink:

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs [hunk 10]
```diff
@@ -25,7 +26,13 @@ internal struct CredHandle
 
             public bool IsZero
             {
-                get { return dwLower == IntPtr.Zero && dwUpper == IntPtr.Zero; }
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                get
+                {
+                    // Ternary operator returning true/false prevents redundant asm generation:
+                    // https://github.com/dotnet/coreclr/issues/914
+                    return dwLower == IntPtr.Zero && dwUpper == IntPtr.Zero ? true : false;
```
Comment (davidsh) [inline src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: This change seems like a temporary workaround to the CLR or JIT. It is not obvious at all to most developers that they need to use a ternary operator. In fact, we have had many PRs over the years where code like this change above has been changed to the original, shortened form `return dwLower == IntPtr.Zero && dwUpper == IntPtr.Zero;`

I don't see the benefit of this change other than an unproven small micro-perf benefit. I would rather see the CLR and related compiler/JIT fixed to make the original pattern performant. This change further reduces readability.

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs [hunk 11]
```diff
@@ -33,9 +34,12 @@ protected SafeDeleteContext() : base(IntPtr.Zero, true)
...
             get
             {
-                return IsClosed || _handle.IsZero;
+                // Ternary operator returning true/false prevents redundant asm generation:
+                // https://github.com/dotnet/coreclr/issues/914
+                return IsClosed || _handle.IsZero ? true : false;
```
Comment (davidsh) [inline src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: See comment above regarding a similar change to this ternary pattern.

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 1 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: The reason chain is
```
Inlines into 0600025A SslStream:get_IsAuthenticated():bool:this
  [0 IL=0014 TR=000011 06000324] [FAILED: unprofitable inline] SecureChannel:get_IsValidContext():bool:this
```
  
```
Inlines into 06000324 SecureChannel:get_IsValidContext():bool:this
  [0 IL=0014 TR=000011 0600041E] [FAILED: unprofitable inline devirt] SafeDeleteContext:get_IsInvalid():bool:this
```

```
Inlines into 0600041E SafeDeleteContext:get_IsInvalid():bool:this
  [1 IL=0001 TR=000001 0600327E] [below ALWAYS_INLINE size] SafeHandle:get_IsClosed():bool:this
  [0 IL=0014 TR=000014 0600001A] [FAILED: unprofitable inline] CredHandle:get_IsZero():bool:this
```

```
Inlines into 0600001A CredHandle:get_IsZero():bool:this
  [1 IL=0011 TR=000003 06001386] [profitable inline] IntPtr:op_Equality(long,long):bool
  [2 IL=0029 TR=000018 06001386] [profitable inline] IntPtr:op_Equality(long,long):bool
```

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 10 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: > This change seems like a temporary workaround to the CLR or JIT.

4.25 years, but I'm not counting https://github.com/dotnet/coreclr/issues/914 😉 

> I don't see the benefit of this change other than an unproven small micro-perf benefit.

It drops 10 asm instructions

```diff  
-   setz al               
-   movzx eax, al         
-   jmp L0039             
-   xor eax, eax          
-   jmp L0040             
-   mov eax, 0x1          
-   test eax, eax         
-   setz al               
-   movzx eax, al         
-   test eax, eax   
```

>  I would rather see the CLR and related compiler/JIT fixed to make the original pattern performant.

Me too; will back out and raise more general issue.

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 1 above)
Comment (AndyAyersMS) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: It's likely the extra apparent complexity from the calls to `operator==`. 

The JIT's profitability assessment for inlining `IsZero` is entirely based on examination of the IL stream. If there are calls in the IL stream the jit has no idea what methods might be called, and in general it is too expensive (at jit time) to ask the runtime to resolve the tokens for calls that an prospective inline candidate method makes. 

I don't expect this state of affairs to change anytime soon, but it is possible.
* I have explored updating the IL based heuristics and remain convinced they can be improved, but it is not going to be easy. 
* There is some hope that with tiering the jit can afford to do a bit more in-depth exploration of inline candidates at Tier1. I don't think we are there quite yet. 
* Perhaps the jit can spend more time doing analysis when prejitting and leave hints to itself that it can consume when jitting.

But for now, in cases like this where those next-level calls are "known" by the programmer to produce compact code, some hinting is needed to get `IsZero` to inline.

File: src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs (hunk 2 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs]: Well its a 4 function chain so is  96,000 x 4 = 384,000 calls; or 12 additional method calls per read and write; vs 4 with the inlined version.

Call counting may be skewing the results; since the JIT doesn't optimize the inlined bools very well without the ternary https://github.com/dotnet/corefx/pull/40265#discussion_r313679833; however each of the method is fairly minimal and it does give the JIT the opportunity to optimize the bools better.

File: src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs (hunk 10 above)
Comment (benaadams) [reply src/Common/src/Interop/Windows/SspiCli/Interop.SSPI.cs]: https://github.com/dotnet/coreclr/issues/914#issuecomment-521083983

File: src/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: Ok.  We can fix it separately.

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks.
